### PR TITLE
Port from 5.0: Fix relro, now, and PIE for Libraries

### PR DESCRIPTION
**Pull request by ivdiazsa**
# Port Description

This is a direct port from one part of PR 685 in the runtime repo. This one was added during .NET Core 5.0 development but there is a customer need to also have it in the .NET Core 3.1 release. The other part is in the core-setup repo, PR 9125.

This change adds the missing `-z,relro` and `-z,now` compile options, as well as the position independent related code and `-pie` linker option to the Native Unix libraries. Additionally, some operating systems require other flags, which were also added:

- *FreeBSD*: `-fuse-ld=lld`
- *Darwin*: `-Wl`, `-bind_at_load`

# Customer Impact
Having these flags enabled provides the built binaries with an additional layer of security, which has become a necessary requirement for some compliance checks, as well as safer applications overall. This port was requested by teams in Azure.

# Regression
This was not a regression.

# Testing
The [checksec](https://github.com/slimm609/checksec.sh) tool was used to verify the executables and shared objects had been indeed built with `PIE` and `Full RELRO` enabled.

# Risk
The risk of this is pretty low, since it's been well tested in the current `main` branch, as well as .NET 5.0 releases.

### CODE REVIEW

File: src/Native/Unix/CMakeLists.txt [hunk 1]
```diff
@@ -145,6 +148,13 @@ endif(CMAKE_SYSTEM_NAME STREQUAL SunOS)
 #       ./build-native.sh cmakeargs -DCLR_ADDITIONAL_COMPILER_OPTIONS=<...> cmakeargs -DCLR_ADDITIONAL_LINKER_FLAGS=<...>
 #
 if(CLR_CMAKE_PLATFORM_UNIX)
+    if (CMAKE_SYSTEM_NAME STREQUAL Darwin)
+        set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,-bind_at_load")
+    else (CMAKE_SYSTEM_NAME STREQUAL Darwin)
+        add_compile_options($<$<COMPILE_LANGUAGE:ASM>:-Wa,--noexecstack>)
+	set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--build-id=sha1 -Wl,-z,relro,-z,now")
```
Comment (janvorli) [inline src/Native/Unix/CMakeLists.txt]: A nit - wrong indentation

File: src/Native/Unix/CMakeLists.txt [hunk 2]
```diff
@@ -122,8 +126,7 @@ endif(CMAKE_SYSTEM_NAME STREQUAL Darwin)
 if(CMAKE_SYSTEM_NAME STREQUAL FreeBSD)
     set(CLR_CMAKE_PLATFORM_UNIX 1)
     add_definitions(-D_BSD_SOURCE) # required for getline
-    set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -fuse-ld=lld -Xlinker --build-id=sha1")
-    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS}  -fuse-ld=lld -Xlinker --build-id=sha1")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fuse-ld=lld")
```
Comment (janvorli) [inline src/Native/Unix/CMakeLists.txt:130]: I've just noticed we are missing the 
```cmake
set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -fuse-ld=lld")
```
here

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Comment by danmoseley** [comment]
This will need the shiproom template?

**Comment by mangod9** [comment]
yeah @ivdiazsa was going to follow up in the .net core ASK mode channel and create the template.

**Comment by aik-jahoda** [comment]
Is this PR targeting current release? If yes we need approve and Servicing-approved tag.

**Review by jeffschwMSFT — APPROVED** [review]
Approved.  I will take for consideration in 3.1.x

cc @blowdart and @GrabYourPitchforks

**Pull request merged**